core/store/src/trie/update.rs

```diff
@@ -279,7 +245,27 @@ impl TrieUpdate {
 
 impl crate::TrieAccess for TrieUpdate {
     fn get(&self, key: &TrieKey) -> Result<Option<Vec<u8>>, StorageError> {
-        TrieUpdate::get(self, key)
+        let key = key.to_vec();
+        if let Some(key_value) = self.prospective.get(&key) {
+            return Ok(key_value.value.as_ref().map(<Vec<u8>>::clone));
+        } else if let Some(changes_with_trie_key) = self.committed.get(&key) {
+            if let Some(RawStateChange { data, .. }) = changes_with_trie_key.changes.last() {
+                return Ok(data.as_ref().map(<Vec<u8>>::clone));
+            }
+        }
+        self.trie.get(&key)
+    }
+
+    fn get_no_side_effects(&self, orig_key: &TrieKey) -> Result<Option<Vec<u8>>, StorageError> {
+        let key = orig_key.to_vec();
+        if let Some(key_value) = self.prospective.get(&key) {
+            return Ok(key_value.value.as_ref().map(<Vec<u8>>::clone));
+        } else if let Some(changes_with_trie_key) = self.committed.get(&key) {
+            if let Some(RawStateChange { data, .. }) = changes_with_trie_key.changes.last() {
+                return Ok(data.as_ref().map(<Vec<u8>>::clone));
+            }
+        }
```

Comment:
this looks exactly the same as for `get`, consider extracting shared code to a separate method, something like `get_from_updates`

Reply:
Done. I'll hold off on merging this until https://github.com/near/nearcore/issues/12027 comes to a conclusion...